lib/mix/lib/mix/scm/git.ex

```diff
@@ -251,6 +251,11 @@ defmodule Mix.SCM.Git do
     :ok
   end
 
+  defp default_branch() do
+    git!(["--git-dir=.git", "remote", "set-head", "origin", "-a"])
```

Comment:
We should probably run this on the `checkout` callback instead.

Reply:
We need to do it on every update because HEAD can change while the dependency is checked out.